File: typescript/packages/core/tests/hooks-fixture-replay.test.ts
/**
 * Cross-language fixture replay for the hook system (spore-core issue #69).
 * Replays all four shared fixtures under `fixtures/hooks/`, asserting the same
 * outcomes the Rust suite asserts. NEVER modify a fixture.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { hooks, SessionId } from "../src/index.js";

const { StandardHookChain, FunctionHook, CommandHook, HookError, hookContextPayload } = hooks;
type HookDecision = hooks.HookDecision;
type HookContext = hooks.HookContext;

const here = dirname(fileURLToPath(import.meta.url));
const fixtureDir = resolve(here, "../../../../fixtures/hooks");
const loadFixture = <T>(name: string): T =>
  JSON.parse(readFileSync(resolve(fixtureDir, name), "utf8")) as T;

// ── hook_decision_wire.json ───────────────────────────────────────────────
describe("hook_decision_wire fixture", () => {
  interface WireCase {
    name: string;
    json: HookDecision;
  }
  const fixture = loadFixture<{ cases: WireCase[] }>("hook_decision_wire.json");

  for (const c of fixture.cases) {
    it(c.name, () => {
      // Round-trip: deserialize then serialize must be byte-identical to `json`.
      const parsed = JSON.parse(JSON.stringify(c.json)) as HookDecision;
      expect(parsed).toEqual(c.json);
      expect(JSON.parse(JSON.stringify(parsed))).toEqual(c.json);
      // The tag key is `decision`.
      expect(typeof (c.json as { decision: string }).decision).toBe("string");
    });
  }
});

// ── pre_tool_use_mutation.json ────────────────────────────────────────────
describe("pre_tool_use_mutation fixture", () => {
  interface MutationCase {
    name: string;
    tool_name: string;
    tool_input: unknown;
    hook_decisions: HookDecision[];
    expected: { outcome: "continue"; tool_input: unknown } | { outcome: "deny"; reason: string };
  }
  const fixture = loadFixture<{ cases: MutationCase[] }>("pre_tool_use_mutation.json");

  for (const c of fixture.cases) {
    it(c.name, async () => {
      const chain = new StandardHookChain();
      c.hook_decisions.forEach((d, i) => {
        chain.register(new FunctionHook(`h${i}`, ["pre_tool_use"], () => d));
      });
      const ctx: HookContext = {
        event: "pre_tool_use",
        session_id: SessionId.of("sess-1"),
        turn_number: 1,
        tool_name: c.tool_name,
        tool_input: c.tool_input,
      };
      const outcome = await chain.fire(ctx);
      if (c.expected.outcome === "continue") {
        expect(outcome.kind).toBe("continue");
        expect(ctx.tool_input).toEqual(c.expected.tool_input);
      } else {
        expect(outcome).toEqual({ kind: "deny", reason: c.expected.reason });
      }
    });
  }
});

// ── stop_block_basic.json ─────────────────────────────────────────────────
describe("stop_block_basic fixture", () => {
  interface StopCase {
    name: string;
    max_stop_blocks: number;
    hook_decisions: HookDecision[];
    expected: { blocks: number; terminated_by: "continue" | "cap" };
  }
  const fixture = loadFixture<{ cases: StopCase[] }>("stop_block_basic.json");

  for (const c of fixture.cases) {
    it(c.name, async () => {
      // Simulate the harness per-run Stop loop: each entry in `hook_decisions`
      // is the verdict at a successive completion gate. A `block` under the cap
      // consumes one block and continues; once `blocks == max_stop_blocks` the
      // next block is ignored and the loop terminates (`cap`); a `continue`
      // terminates immediately (`continue`).
      let blocks = 0;
      let terminatedBy: "continue" | "cap" = "continue";
      for (const decision of c.hook_decisions) {
        const chain = new StandardHookChain();
        chain.register(new FunctionHook("stop", ["stop"], () => decision));
        const ctx: HookContext = {
          event: "stop",
          session_id: SessionId.of("sess-1"),
          turn_number: blocks,
          last_output: { text: "", had_tool_calls: false },
          task_instruction: "do it",
          session_state: null,
        };
        const outcome = await chain.fire(ctx);
        if (outcome.kind === "block") {
          if (blocks >= c.max_stop_blocks) {
            terminatedBy = "cap";
            break;
          }
          blocks += 1;
          continue;
        }
        // continue / inject / deny → normal termination.
        terminatedBy = "continue";
        break;
      }
      expect(blocks).toBe(c.expected.blocks);
      expect(terminatedBy).toBe(c.expected.terminated_by);
    });
  }
});

// ── command_handler_io.json ───────────────────────────────────────────────
describe("command_handler_io fixture", () => {
  interface CmdCase {
    name: string;
    event: hooks.HookEvent;
    expected_stdin?: { event: string; context: Record<string, unknown> };
    exit_code?: number;
    stdout: string;
    expected_decision?: HookDecision;
    expected_error?: "command_failed" | "command_output_invalid";
  }
  const fixture = loadFixture<{ cases: CmdCase[] }>("command_handler_io.json");

  // Build the live HookContext that matches each fixture case's expected_stdin.
  function ctxFor(c: CmdCase): HookContext {
    if (c.event === "stop") {
      return {
        event: "stop",
        session_id: SessionId.of("sess-1"),
        turn_number: 3,
        last_output: { text: "I'm done", had_tool_calls: false },
        task_instruction: "make the tests pass",
        session_state: null,
      };
    }
    // pre_tool_use
    return {
      event: "pre_tool_use",
      session_id: SessionId.of("sess-1"),
      turn_number: 1,
      tool_name: "read_file",
      tool_input: { path: "/etc/passwd" },
    };
  }

  for (const c of fixture.cases) {
    it(c.name, async () => {
      const ctx = ctxFor(c);

      // 1. The stdin payload the handler WOULD receive matches expected_stdin.
      if (c.expected_stdin) {
        const payload = {
          event: ctx.event,
          context: hookContextPayload(ctx),
        };
        // Compare via JSON normalization (SessionId → string).
        expect(JSON.parse(JSON.stringify(payload))).toEqual(c.expected_stdin);
      }

      // 2. Drive a real CommandHook whose script echoes `stdout` and exits with
      //    `exit_code` (default 0).
      const exit = c.exit_code ?? 0;
      const script = `cat >/dev/null; printf '%s' ${shellQuote(c.stdout)}; exit ${exit}`;
      const hook = new CommandHook("cmd", [c.event], "sh", ["-c", script]);
      const chain = new StandardHookChain();
      chain.register(hook);

      if (c.expected_error) {
        await expect(chain.fire(ctx)).rejects.toMatchObject({ kind: c.expected_error });
        await expect(chain.fire(ctx)).rejects.toBeInstanceOf(HookError);
      } else {
        const outcome = await chain.fire(ctx);
        // The parsed decision must equal expected_decision; map to FireOutcome.
        const d = c.expected_decision!;
        switch (d.decision) {
          case "block":
            expect(outcome).toEqual({ kind: "block", reason: d.reason });
            break;
          case "deny":
            expect(outcome).toEqual({ kind: "deny", reason: d.reason });
            break;
          case "continue":
            expect(outcome).toEqual({ kind: "continue" });
            break;
          default:
            throw new Error(`unexpected fixture decision ${d.decision}`);
        }
      }
    });
  }
});

function shellQuote(s: string): string {
  return `'${s.replace(/'/g, `'\\''`)}'`;
}

File: go/spore-core/hooks_fixture_replay_test.go
package sporecore

import (
	"bytes"
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

// hooksFixturePath resolves a file under fixtures/hooks/.
func hooksFixturePath(t *testing.T, name string) string {
	t.Helper()
	_, this, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	return filepath.Join(filepath.Dir(this), "..", "..", "fixtures", "hooks", name)
}

func readFixture(t *testing.T, name string) []byte {
	t.Helper()
	raw, err := os.ReadFile(hooksFixturePath(t, name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	return raw
}

// ── hook_decision_wire.json ────────────────────────────────────────────────
// Serializing each variant must reproduce `json` byte-for-byte (compacted),
// and deserializing `json` must round-trip.
func TestFixtureHookDecisionWire(t *testing.T) {
	var fixture struct {
		Cases []struct {
			Name string          `json:"name"`
			JSON json.RawMessage `json:"json"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(readFixture(t, "hook_decision_wire.json"), &fixture); err != nil {
		t.Fatal(err)
	}
	if len(fixture.Cases) != 5 {
		t.Fatalf("expected 5 cases, got %d", len(fixture.Cases))
	}
	for _, c := range fixture.Cases {
		t.Run(c.Name, func(t *testing.T) {
			var d HookDecision
			if err := json.Unmarshal(c.JSON, &d); err != nil {
				t.Fatalf("unmarshal: %v", err)
			}
			got, err := json.Marshal(d)
			if err != nil {
				t.Fatal(err)
			}
			want := compact(t, c.JSON)
			if !bytes.Equal(got, want) {
				t.Fatalf("got %s, want %s", got, want)
			}
		})
	}
}

// ── command_handler_io.json ─────────────────────────────────────────────────
// Pins the CommandHook stdin shape (event + per-event context) and the parsed
// decision / error for each case.
func TestFixtureCommandHandlerIO(t *testing.T) {
	var fixture struct {
		Cases []struct {
			Name             string          `json:"name"`
			Event            HookEvent       `json:"event"`
			ExpectedStdin    json.RawMessage `json:"expected_stdin"`
			Stdout           string          `json:"stdout"`
			ExitCode         int             `json:"exit_code"`
			ExpectedDecision json.RawMessage `json:"expected_decision"`
			ExpectedError    string          `json:"expected_error"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(readFixture(t, "command_handler_io.json"), &fixture); err != nil {
		t.Fatal(err)
	}

	for _, c := range fixture.Cases {
		t.Run(c.Name, func(t *testing.T) {
			hctx := stopContextFromFixtureEvent(t, c.Event)

			// 1. Verify the stdin payload the harness would send matches
			//    expected_stdin (when the fixture pins it).
			if len(c.ExpectedStdin) > 0 {
				payload, err := hctx.toPayload()
				if err != nil {
					t.Fatal(err)
				}
				stdin, err := json.Marshal(struct {
					Event   HookEvent       `json:"event"`
					Context json.RawMessage `json:"context"`
				}{c.Event, payload})
				if err != nil {
					t.Fatal(err)
				}
				// Compare semantically: the harness emits map keys in sorted
				// order, while the fixture pins an author-chosen key order. JSON
				// object key order is not significant on the wire.
				if !jsonEqual(t, stdin, c.ExpectedStdin) {
					t.Fatalf("stdin\n got: %s\nwant: %s", stdin, compact(t, c.ExpectedStdin))
				}
			}

			// 2. Drive a real CommandHook whose script emits the fixture's
			//    stdout and exit code, and assert the parsed decision / error.
			hook := scriptHook(t, c.Event, c.Stdout, c.ExitCode)
			chain := NewStandardHookChain()
			if err := chain.Register(hook); err != nil {
				t.Fatal(err)
			}
			outcome, err := chain.Fire(context.Background(), hctx)

			switch c.ExpectedError {
			case "":
				if err != nil {
					t.Fatalf("unexpected error: %v", err)
				}
				var want HookDecision
				if err := json.Unmarshal(c.ExpectedDecision, &want); err != nil {
					t.Fatal(err)
				}
				assertOutcomeMatchesDecision(t, outcome, want)
			case "command_failed":
				var he *HookError
				if !asHookError(err, &he) || he.Kind != HookErrCommandFailed {
					t.Fatalf("err = %v, want CommandFailed", err)
				}
			case "command_output_invalid":
				var he *HookError
				if !asHookError(err, &he) || he.Kind != HookErrCommandOutputInvalid {
					t.Fatalf("err = %v, want CommandOutputInvalid", err)
				}
			default:
				t.Fatalf("unknown expected_error %q", c.ExpectedError)
			}
		})
	}
}

// stopContextFromFixtureEvent builds the HookContext matching the fixture's
// pinned stdin for the given event.
func stopContextFromFixtureEvent(t *testing.T, event HookEvent) *HookContext {
	t.Helper()
	switch event {
	case HookEventStop:
		out := TurnOutput{Text: "I'm done", HadToolCalls: false}
		instr := "make the tests pass"
		return &HookContext{
			Event:           HookEventStop,
			SessionID:       SessionID("sess-1"),
			TurnNumber:      3,
			LastOutput:      &out,
			TaskInstruction: &instr,
			SessionState:    nil, // fixture pins session_state: null
		}
	case HookEventPreToolUse:
		input := json.RawMessage(`{"path":"/etc/passwd"}`)
		return &HookContext{
			Event:      HookEventPreToolUse,
			SessionID:  SessionID("sess-1"),
			TurnNumber: 1,
			ToolName:   "read_file",
			ToolInput:  &input,
		}
	default:
		t.Fatalf("unhandled fixture event %q", event)
		return nil
	}
}

// scriptHook builds a CommandHook backed by a shell script that consumes stdin,
// emits stdout, and exits with exitCode.
func scriptHook(t *testing.T, event HookEvent, stdout string, exitCode int) *CommandHook {
	t.Helper()
	script := filepath.Join(t.TempDir(), "hook.sh")
	body := "#!/bin/sh\ncat >/dev/null\n"
	if stdout != "" {
		body += "printf '%s' " + shellQuote(stdout) + "\n"
	}
	body += "exit " + itoaFull(exitCode) + "\n"
	if err := os.WriteFile(script, []byte(body), 0o755); err != nil {
		t.Fatal(err)
	}
	return NewCommandHook("fixture", []HookEvent{event}, "sh", []string{script})
}

func assertOutcomeMatchesDecision(t *testing.T, outcome FireOutcome, want HookDecision) {
	t.Helper()
	switch want.Decision {
	case HookDecisionBlock:
		if outcome.Kind != FireBlock || outcome.Reason != want.Reason {
			t.Fatalf("outcome = %+v, want block %q", outcome, want.Reason)
		}
	case HookDecisionDeny:
		if outcome.Kind != FireDeny || outcome.Reason != want.Reason {
			t.Fatalf("outcome = %+v, want deny %q", outcome, want.Reason)
		}
	case HookDecisionContinue:
		if outcome.Kind != FireContinue {
			t.Fatalf("outcome = %+v, want continue", outcome)
		}
	default:
		t.Fatalf("unhandled expected decision %q", want.Decision)
	}
}

// ── pre_tool_use_mutation.json ──────────────────────────────────────────────
// Replays an ordered chain of decisions against a tool_input and asserts the
// final (mutated) input or a deny.
func TestFixturePreToolUseMutation(t *testing.T) {
	var fixture struct {
		Cases []struct {
			Name          string            `json:"name"`
			ToolName      string            `json:"tool_name"`
			ToolInput     json.RawMessage   `json:"tool_input"`
			HookDecisions []json.RawMessage `json:"hook_decisions"`
			Expected      struct {
				Outcome   string          `json:"outcome"`
				ToolInput json.RawMessage `json:"tool_input"`
				Reason    string          `json:"reason"`
			} `json:"expected"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(readFixture(t, "pre_tool_use_mutation.json"), &fixture); err != nil {
		t.Fatal(err)
	}

	for _, c := range fixture.Cases {
		t.Run(c.Name, func(t *testing.T) {
			chain := NewStandardHookChain()
			for i, raw := range c.HookDecisions {
				var d HookDecision
				if err := json.Unmarshal(raw, &d); err != nil {
					t.Fatal(err)
				}
				name := "h" + itoaFull(i)
				if err := chain.Register(fnHook(name, []HookEvent{HookEventPreToolUse}, d)); err != nil {
					t.Fatal(err)
				}
			}
			input := append(json.RawMessage(nil), compact(t, c.ToolInput)...)
			hctx := &HookContext{
				Event:      HookEventPreToolUse,
				SessionID:  SessionID("sess-1"),
				TurnNumber: 1,
				ToolName:   c.ToolName,
				ToolInput:  &input,
			}
			outcome, err := chain.Fire(context.Background(), hctx)
			if err != nil {
				t.Fatal(err)
			}
			switch c.Expected.Outcome {
			case "continue":
				if outcome.Kind != FireContinue {
					t.Fatalf("outcome = %+v, want continue", outcome)
				}
				if !jsonEqual(t, input, c.Expected.ToolInput) {
					t.Fatalf("tool_input = %s, want %s", input, compact(t, c.Expected.ToolInput))
				}
			case "deny":
				if outcome.Kind != FireDeny || outcome.Reason != c.Expected.Reason {
					t.Fatalf("outcome = %+v, want deny %q", outcome, c.Expected.Reason)
				}
			default:
				t.Fatalf("unknown outcome %q", c.Expected.Outcome)
			}
		})
	}
}

// ── stop_block_basic.json ───────────────────────────────────────────────────
// Replays a sequence of Stop-hook decisions through the per-run block-count /
// cap logic (mirroring StandardHarness.fireStopHooks) and asserts the honored
// block count and how termination was reached.
func TestFixtureStopBlockBasic(t *testing.T) {
	var fixture struct {
		Cases []struct {
			Name          string            `json:"name"`
			MaxStopBlocks uint32            `json:"max_stop_blocks"`
			HookDecisions []json.RawMessage `json:"hook_decisions"`
			Expected      struct {
				Blocks       int    `json:"blocks"`
				TerminatedBy string `json:"terminated_by"`
			} `json:"expected"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(readFixture(t, "stop_block_basic.json"), &fixture); err != nil {
		t.Fatal(err)
	}

	for _, c := range fixture.Cases {
		t.Run(c.Name, func(t *testing.T) {
			// Decode the scripted decisions, fired one per simulated termination
			// attempt.
			decisions := make([]HookDecision, len(c.HookDecisions))
			for i, raw := range c.HookDecisions {
				if err := json.Unmarshal(raw, &decisions[i]); err != nil {
					t.Fatal(err)
				}
			}
			cap := c.MaxStopBlocks

			var blocks uint32
			terminatedBy := ""
			// Walk the decision sequence: each is the Stop-chain's outcome at
			// one termination attempt. A block under the cap consumes one block
			// and continues; once blocks == cap the next block is ignored and
			// the loop terminates ("cap"); a continue terminates ("continue").
			for _, d := range decisions {
				if d.Decision == HookDecisionBlock {
					if blocks >= cap {
						terminatedBy = "cap"
						break
					}
					blocks++
					continue
				}
				// Continue (or any non-block) → normal termination.
				terminatedBy = "continue"
				break
			}
			if terminatedBy == "" {
				// Ran out of scripted decisions while still blocking under cap;
				// treat the final state as cap-bound termination.
				terminatedBy = "cap"
			}

			if int(blocks) != c.Expected.Blocks {
				t.Fatalf("blocks = %d, want %d", blocks, c.Expected.Blocks)
			}
			if terminatedBy != c.Expected.TerminatedBy {
				t.Fatalf("terminated_by = %q, want %q", terminatedBy, c.Expected.TerminatedBy)
			}
		})
	}
}

// jsonEqual reports whether two JSON documents are semantically equal
// (key order and insignificant whitespace ignored).
func jsonEqual(t *testing.T, a, b json.RawMessage) bool {
	t.Helper()
	var av, bv any
	if err := json.Unmarshal(a, &av); err != nil {
		t.Fatalf("unmarshal a: %v", err)
	}
	if err := json.Unmarshal(b, &bv); err != nil {
		t.Fatalf("unmarshal b: %v", err)
	}
	ab, _ := json.Marshal(av)
	bb, _ := json.Marshal(bv)
	return bytes.Equal(ab, bb)
}

// compact returns the compacted (canonical) form of a JSON document.
func compact(t *testing.T, raw json.RawMessage) []byte {
	t.Helper()
	var buf bytes.Buffer
	if err := json.Compact(&buf, raw); err != nil {
		t.Fatalf("compact: %v", err)
	}
	return buf.Bytes()
}

func itoaFull(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b []byte
	for n > 0 {
		b = append([]byte{byte('0' + n%10)}, b...)
		n /= 10
	}
	if neg {
		b = append([]byte{'-'}, b...)
	}
	return string(b)
}

func shellQuote(s string) string {
	return "'" + bytesReplaceAll(s, "'", `'\''`) + "'"
}

func bytesReplaceAll(s, old, new string) string {
	return string(bytes.ReplaceAll([]byte(s), []byte(old), []byte(new)))
}
